ur_controllers/src/scaled_joint_trajectory_controller.cpp

```diff
@@ -19,10 +19,212 @@
+{
+controller_interface::InterfaceConfiguration ScaledJointTrajectoryController::state_interface_configuration() const
+{
+  controller_interface::InterfaceConfiguration conf;
+  conf = JointTrajectoryController::state_interface_configuration();
+  conf.names.push_back("speed_scaling/speed_scaling_factor");
+  return conf;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ScaledJointTrajectoryController::on_activate(const rclcpp_lifecycle::State& state)
+{
+  TimeData time_data;
+  time_data.time = node_->now();
+  time_data.period = rclcpp::Duration(0, 0);
+  time_data.uptime = node_->now();
+  time_data_.initRT(time_data);
+  return JointTrajectoryController::on_activate(state);
+}
+
+controller_interface::return_type ScaledJointTrajectoryController::update()
+{
+  if (state_interfaces_.back().get_name() == "speed_scaling")
+  {
+    scaling_factor_ = state_interfaces_.back().get_value();
+  }
+  else
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "Speed scaling interface not fount. This should not have happended");
+  }
+
+  if (get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE)
+  {
+    if (!is_halted)
+    {
+      halt();
+      is_halted = true;
+    }
+    return controller_interface::return_type::SUCCESS;
+  }
+
+  auto resize_joint_trajectory_point = [](trajectory_msgs::msg::JointTrajectoryPoint& point, size_t size) {
+    point.positions.resize(size);
+    point.velocities.resize(size);
+    point.accelerations.resize(size);
+  };
+  auto compute_error_for_joint = [](JointTrajectoryPoint& error, int index, const JointTrajectoryPoint& current,
+                                    const JointTrajectoryPoint& desired) {
+    // error defined as the difference between current and desired
+    error.positions[index] = angles::shortest_angular_distance(current.positions[index], desired.positions[index]);
+    error.velocities[index] = desired.velocities[index] - current.velocities[index];
+    error.accelerations[index] = 0.0;
+  };
+
+  // Check if a new external message has been received from nonRT threads
+  auto current_external_msg = traj_external_point_ptr_->get_trajectory_msg();
+  auto new_external_msg = traj_msg_external_point_ptr_.readFromRT();
+  if (current_external_msg != *new_external_msg)
+  {
+    fill_partial_goal(*new_external_msg);
+    sort_to_local_joint_order(*new_external_msg);
+    traj_external_point_ptr_->update(*new_external_msg);
+  }
+
+  JointTrajectoryPoint state_current, state_desired, state_error;
+  const auto joint_num = joint_names_.size();
+  resize_joint_trajectory_point(state_current, joint_num);
+
+  // current state update
+  for (auto index = 0ul; index < joint_num; ++index)
+  {
+    state_current.positions[index] = joint_position_state_interface_[index].get().get_value();
+    state_current.velocities[index] = joint_velocity_state_interface_[index].get().get_value();
+    state_current.accelerations[index] = 0.0;
```

Comment:
Hardcoded value is ok?

Reply:
This was copied from the JTC. I think the acceleration is currently not set. Was the same in ROS1 if I remember correctly